WordPress/gutenberg
# Add table block keyboard navigation

Status: This is blocked by https://github.com/WordPress/gutenberg/pull/17084, which will move some of the functionality from this PR to a `WritingFlow`.

## Description
This PR implements improved keyboard navigation within the table block. Previously only tab could be used to navigate cells.

This PR implements the following changes:
- Arrow keys are now used to navigate cells along with `HOME` and `END`. The `Cmd`/`Ctrl` modifiers can be used to go directly to the start/end of a column, row, or the entire table (when used with `HOME` and `END`)
![table-arrow-nav](https://user-images.githubusercontent.com/677833/62997097-a17b5900-be99-11e9-9c8c-ee19a465f761.gif)

- Tab / Shift+Tab is now used to navigate directly to from an individual cell within the table. A user is able to navigate to a cell using arrow keys, shift tab to the toolbar to make some changes to the table, then tab back to the last selected cell. Previously this was quite hard to achieve as tabbing to the toolbar also changed the selected cell.
![table-tab](https://user-images.githubusercontent.com/677833/62997102-a4764980-be99-11e9-96b0-c3f48e6fdc6e.gif)

The implementation follows the described behaviours here and uses a roving tabIndex:
https://www.w3.org/TR/wai-aria-practices/examples/grid/dataGrids.html

The changes here open up the possibilities for other future enhancements to the table block:
- Multi-cell selection using the keyboard (holding down shift with navigation keys).
- Merging of cells (`colspan`/`rowspan`), which will require multi-selection.
- The improvements to keyboard usage also help considerably with implementing column/row selection (#16493), which is why I switched to working on this first.

## How has this been tested?
Added e2e and unit tests

Manual testing:
1. Add a table block
2. Toggle on the header and footer.
3. Use cursor keys to navigate around. Ensure all cells are navigable.
4. Type some text.
5. Use cursor keys to navigate through the cell with text.
6. Use HOME and END to navigate to the start and end of the row.
7. Use Cmd/Ctrl with arrow keys and HOME and END
8. Use shift tab and tab to navigate out of the table and back in to the last selected cell.

## Types of changes


New feature (non-breaking change which adds functionality)


## Checklist:
